### ISSUE TITLE
trace: don't finish spans twice when they are propagated

### ISSUE THREAD

**Issue by rakyll** [GoogleCloudPlatform/google-cloud-go#532]
Consider the following example where the span tracing the https://users.service/users request will be finished by fetchUsers function:

``` go
func fetchUsers() ([]*User, error) {
	span := traceClient.NewSpan("/users")
	defer span.Finish()

	// Create the outgoing request, a GET to the users endpoint.
	req, _ := http.NewRequest("GET", "https://users.service/users", nil)

	// Create a span to attach tracing information.
	rspan := span.NewRemoteChild(req)
	defer rspan.Finish()

	// do the request, etc.
}
```

On the users.service side, our APIs allow the propagated span to be finished by the remote service:

``` go
func usersHandler(w http.ResponseWriter, r *http.Request) {
	span := traceClient.SpanFromRequest(r)
        defer span.Finish() // NOT THE RESPONSIBILITY OF THIS SERVICE TO FINISH THIS SPAN!!

	req, _ := http.NewRequest("GET", "https://users.service/meta", nil)

	rspan := span.NewRemoteChild(req)
	defer rspan.Finish()

	// ...
}
```

In the new [trace prototype](https://godoc.org/github.com/rakyll/trace) we have been working on, we came up with the solution of introducing FinishFuncs. Each span creator is returning a FinishFunc if the span is finish-able:

``` go
ctx, finish := trace.ChildSpan(ctx, "/users")
defer finish()
```

If you propagate and resurrect a span in the remote server, we are not giving you a finisher. So it is explicit that it is not responsibility to finish the span:

``` go
func Span(ctx context.Context, name string, info []byte) (context.Context, error)
```

We should follow a similar approach here and make sure that our APIs restrict the users not to misuse the finish feature.

**Comment by rakyll** [comment]
Blocked by the behavior of SpanFromRequest, see https://github.com/GoogleCloudPlatform/google-cloud-go/issues/538.

**Comment by rakyll** [comment]
Closing in favor of https://github.com/GoogleCloudPlatform/google-cloud-go/issues/549.
